src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -233,19 +236,29 @@ public void onFailure(Throwable e) {
             if (ExceptionsHelper.unwrapCause(e) instanceof ConnectTransportException) {
                 int i = ++this.i;
                 if (i >= nodes.size()) {
-                    listener.onFailure(new NoNodeAvailableException("None of the configured nodes were available: " + nodes, e));
+                    runFailureInGenericThreadPool(new NoNodeAvailableException("None of the configured nodes were available: " + nodes, e));
                 } else {
                     try {
                         callback.doWithNode(nodes.get((index + i) % nodes.size()), this);
-                    } catch(Throwable t) {
-                        //this exception can't come from the TransportService as it doesn't throw exceptions at all
-                        listener.onFailure(t);
+                    } catch(final Throwable t) {
+                        // this exception can't come from the TransportService as it doesn't throw exceptions at all
+                        runFailureInGenericThreadPool(t);
                     }
                 }
             } else {
-                listener.onFailure(e);
+                runFailureInGenericThreadPool(e);
             }
         }
+
+        // need to ensure to not block the netty I/O thread, in case of retry due to the node sampling
+        private void runFailureInGenericThreadPool(final Throwable t) {
```

Comment:
I was unsure of doing this, due to checking the `ThreadPool` constructor, showing that the `LISTENER` threadpool does not have any queue